pkg/source/openstack/client.go

```diff
@@ -563,3 +595,41 @@ func mapNetworkCards(networkCards []networkInfo, mapping []migration.NetworkMapp
 
 	return retNetwork
 }
+
+func (c *Client) ImageFirmwareSettings(instance *servers.Server) (bool, bool, bool, error) {
+	var imageID string
+	var uefiType, tpmEnabled, secureBoot bool
+	for _, v := range instance.AttachedVolumes {
+		resp := volumes.Get(c.storageClient, v.ID)
+		var volInfo volumes.Volume
+		if err := resp.ExtractIntoStructPtr(&volInfo, "volume"); err != nil {
+			return uefiType, tpmEnabled, secureBoot, fmt.Errorf("error extracting volume info for volume %s: %v", v.ID, err)
+		}
+
+		if volInfo.Bootable == "true" {
+			var volStatus ExtendedVolume
+			if err := resp.ExtractIntoStructPtr(&volStatus, "volume"); err != nil {
+				return uefiType, tpmEnabled, secureBoot, fmt.Errorf("error extracting volume status for volume %s: %v", v.ID, err)
+			}
+			imageID = volStatus.VolumeImageMetadata["image_id"]
+		}
+	}
+
+	imageInfo, err := images.Get(c.imageClient, imageID).Extract()
+	if err != nil {
+		return uefiType, tpmEnabled, secureBoot, fmt.Errorf("error getting image details for image %s: %v", imageID, err)
+	}
+	firmwareType, ok := imageInfo.Properties["hw_firmware_type"]
+	if ok && firmwareType.(string) == "uefi" {
+		uefiType = true
+	}
+	logrus.Info(imageInfo.Properties)
```

Comment:
Do we need to add more messages which we can identify where it's from when debugging?

Reply:
done.